pkg/utils/set/set.go

```diff
@@ -0,0 +1,118 @@
+package set
+
+import (
+	"fmt"
+)
+
+// Set is a generic thread-safe data structure representing a unique collection of items
+type Set[T, HashType comparable] struct {
+	items     []T
+	uniqueSet map[HashType]T
+	hashFunc  func(T) HashType
+}
+
+// New builds a new set object
+func New[T comparable](items ...T) *Set[T, T] {
+	set := Set[T, T]{
+		items:     make([]T, 0, len(items)),
+		uniqueSet: make(map[T]T, len(items)),
+		hashFunc: func(t T) T {
+			return t
+		},
+	}
+	for _, item := range items {
+		set.uniqueSet[item] = item
+	}
+	for item := range set.uniqueSet {
+		set.items = append(set.items, item)
+	}
+	return &set
+}
+
+// New builds a new set object
+func NewWithHash[T comparable, HashType comparable](hashFunc func(T) HashType, items ...T) *Set[T, HashType] {
```

Comment:
Looks like too much for this type (as its already a set). But since you are already implementing the ThreadSafeSet type, if we created a HashSet type we would also have to create a ThreadSafeHashSet. IMO it would be better to distinguish both, but I dont think it really matters (since you have a constructor that creates a default basic hash).